collector/export/metric_otlp.go

```diff
@@ -0,0 +1,299 @@
+			Close:                 opts.Close,
+			MaxIdleConnsPerHost:   opts.MaxIdleConnsPerHost,
+			MaxIdleConns:          opts.MaxIdleConns,
+			IdleConnTimeout:       opts.IdleConnTimeout,
+			ResponseHeaderTimeout: opts.ResponseHeaderTimeout,
+			DisableHTTP2:          opts.DisableHTTP2,
+			DisableKeepAlives:     opts.DisableKeepAlives,
+		},
+	)
+
+	var attributes []*commonv1.KeyValue
+	if opts.AddResourceAttributes != nil {
+		attributes = make([]*commonv1.KeyValue, 0, len(opts.AddResourceAttributes))
+		for k, v := range opts.AddResourceAttributes {
+			attributes = append(attributes, &commonv1.KeyValue{
+				Key:   k,
+				Value: &commonv1.AnyValue{Value: &commonv1.AnyValue_StringValue{StringValue: v}},
+			})
+		}
+	}
+	return &PromToOtlpExporter{
+		httpClient:         client,
+		destination:        opts.Destination,
+		transformer:        opts.Transformer,
+		resourceAttributes: attributes,
+
+		stringKVPool: pool.NewGeneric(4096, func(sz int) interface{} {
+			return &commonv1.KeyValue{
+				Key:   "",
+				Value: &commonv1.AnyValue{Value: &commonv1.AnyValue_StringValue{StringValue: ""}},
+			}
+		}),
+
+		datapointPool: pool.NewGeneric(4096, func(sz int) interface{} {
+			return &metricsv1.NumberDataPoint{
+				Attributes:   nil,
+				TimeUnixNano: 0,
+				Value:        &metricsv1.NumberDataPoint_AsDouble{AsDouble: 0},
+			}
+		}),
+	}
+}
+
+// Write sends the given WriteRequest to the given OTLP endpoint
+func (c *PromToOtlpExporter) Write(ctx context.Context, wr *prompb.WriteRequest) error {
+	wr = c.transformer.TransformWriteRequest(wr)
+	count := int64(len(wr.Timeseries))
+	if count == 0 {
+		return nil
+	}
+
+	serialized, err := c.promToOtlpRequest(wr)
+	if err != nil {
+		return fmt.Errorf("metric otlp forwarder convert: %w", err)
+	}
+
+	return c.sendRequest(serialized, count)
+}
+
+func (c *PromToOtlpExporter) CloseIdleConnections() {
+	c.httpClient.CloseIdleConnections()
+}
+
+func (c *PromToOtlpExporter) promToOtlpRequest(wr *prompb.WriteRequest) ([]byte, error) {
+	scopeMetrics := &metricsv1.ScopeMetrics{
+		Metrics: make([]*metricsv1.Metric, 0, len(wr.Timeseries)),
+	}
+	exportRequest := &v1.ExportMetricsServiceRequest{
+		ResourceMetrics: []*metricsv1.ResourceMetrics{
+			{
+				Resource: &resourcev1.Resource{
+					Attributes: c.resourceAttributes,
+				},
+				ScopeMetrics: []*metricsv1.ScopeMetrics{
+					scopeMetrics,
+				},
+			},
+		},
+	}
+
+	for _, ts := range wr.Timeseries {
+		gauge := &metricsv1.Gauge{
+			DataPoints: make([]*metricsv1.NumberDataPoint, 0, len(ts.Samples)),
+		}
+		metric := &metricsv1.Metric{
+			Data: &metricsv1.Metric_Gauge{
+				Gauge: gauge,
+			},
+		}
+
+		var nameBytes []byte
+		attributes := make([]*commonv1.KeyValue, 0, len(ts.Labels))
+		for _, element := range ts.Labels {
+			if bytes.Equal(element.Name, nameLabel) {
+				nameBytes = element.Value
+			} else if !bytes.HasPrefix(element.Name, []byte("adxmon_")) {
+				attribute := c.stringKVPool.Get(0).(*commonv1.KeyValue)
```

Comment:
Does attribute need to be reset after getting it from the pool?  Wondering if there is any chance for prior used data to still be present.